feat(proto): add physical hashJoin & mergeJoin physical relations (#336)
substrait-io/substrait: site/docs/relations/physical_relations.md

| Property            | Description                                                                                                                                                                                                                                                                                     | Required                                                                   |
|---------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------|
| Left Input          | A relational input.                                                                                                                                                                                                                                                                             | Required                                                                   |

> | Right Input         | A relational input.                                                                                                                                                                                                                                                                             | Required                                                                   |

Reviewer: Since this is a physical relation do we want to explicitly declare that the left side is the build side and the right side is the probe side (or vice versa)?  It looks like you are already making this assumption in your description of left keys and right keys.

Author: I think we can leave it to the user to decide which side is probe or build , your thoughts?

Reviewer: Picking the correct side is generally an optimizer task (requires statistics / heuristics).  My assumption for more physical nodes is that they've already gone through an optimizer.

For a single optimizer/consumer pair this wouldn't be a problem.  They can both agree offline which side should be build and probe.  However, imagine an engine-independent substrait optimizer created this plan and hoped for it to be consumed by many different engines.  In that case the different consumers might pick left/right build/probe differently.

That being said, I haven't worked much with optimizers.  CC @jacques-n who has more experience here.

>  | Left Keys           | A collection of fields reference in probe table                                                                                                                                                                                                                                                 | Optional, defaults to blank if during join predicate is specified          |

Reviewer: ```suggestion
 | Left Keys           | References to the fields to join on in the left input.                                                                                                                                                                                                                                                 | Optional, defaults to blank if during join predicate is specified          |
```

Same change for right keys

 | Right Keys          | A collection of fields reference in build table                                                                                                                                                                                                                                                 | Optional, defaults to blank if during join predicate is specified          |             

> | During Join Predicate | A boolean condition that describes whether each record from the left set "match" the record from the right set. The condition must only include the following operations: AND, ==, field references, is not distinct from. Field references correspond to the direct output order of the data.  | Optional, defaults true if left keys and right keys are both specified.    |

Reviewer: I don't see this in the proto definition

Author: Removed

### Hash Equijoin Properties

| Property            | Description                                                                                                                                                                                                             | Required                 |
|---------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------|
| Left Input          | A relational input.                                                                                                                                                                                                     | Required                 |

> | Right Input         | A relational input.                                                                                                                                                                                                     | Required                 |

Reviewer: Either we should assign these as probe and build (my preference would be to be explicit) or we should specifically explain that consumers should assign probe & build based on whatever criteria they want.  As it is, I think this is ambiguous.

Author: I found that the documentation has mentioned that the right input is build-side, see [this](https://substrait.io/relations/physical_relations/#hash-equijoin-operator)

I also updated in this section [https://substrait.io/relations/physical_relations/#hash-equijoin-properties]
